packages/smithy-core/src/smithy_core/aio/client.py

```diff
@@ -0,0 +1,581 @@
+class ClientCall[I: SerializeableShape, O: DeserializeableShape]:
+    """A data class containing all the initial information about an operation
+    invocation."""
+
+    input: I
+    """The input of the operation."""
+
+    operation: APIOperation[I, O] = field(repr=False)
+    """The schema of the operation."""
+
+    context: TypedProperties
+    """The initial context of the operation."""
+
+    interceptor: Interceptor[I, O, Any, Any]
+    """The interceptor to use in the course of the operation invocation.
+
+    This SHOULD be an InterceptorChain.
+    """
+
+    auth_scheme_resolver: AuthSchemeResolver
+    """The auth scheme resolver for the operation."""
+
+    supported_auth_schemes: dict[ShapeID, AuthScheme[Any, Any, Any, Any]]
+    """The supported auth schemes for the operation."""
+
+    endpoint_resolver: EndpointResolver
+    """The endpoint resolver for the operation."""
+
+    retry_strategy: RetryStrategy
+    """The retry strategy to use for the operation."""
+
+    retry_scope: str | None = None
+    """The retry scope for the operation."""
+
+    def retryable(self) -> bool:
+        # TODO: check to see if the stream is seekable
+        return self.operation.input_stream_member is None
+
+
+class RequestPipeline[TRequest: Request, TResponse: Response]:
+    """Invokes client operations asynchronously."""
+
+    protocol: ClientProtocol[TRequest, TResponse]
+    """The protocol to use to serialize the request and deserialize the response."""
+
+    transport: ClientTransport[TRequest, TResponse]
+    """The transport to use to send the request and receive the response (e.g. an HTTP
+    Client)."""
+
+    def __init__(
+        self,
+        protocol: ClientProtocol[TRequest, TResponse],
+        transport: ClientTransport[TRequest, TResponse],
+    ) -> None:
+        self.protocol = protocol
+        self.transport = transport
+
+    async def __call__[I: SerializeableShape, O: DeserializeableShape](
+        self, call: ClientCall[I, O], /
+    ) -> O:
+        """Invoke an operation asynchronously.
+
+        :param call: The operation to invoke and associated context.
+        """
+        output, _ = await self._execute_request(call, None)
+        return output
+
+    async def input_stream[
+        I: SerializeableShape,
+        O: DeserializeableShape,
+        E: SerializeableShape,
+    ](
+        self, call: ClientCall[I, O], event_type: "TypeForm[E]", /
+    ) -> InputEventStream[E, O]:
+        """Invoke an input stream operation asynchronously.
+
+        :param call: The operation to invoke and associated context.
+        :param event_type: The event type to send in the input stream.
+        """
+        request_future = Future[RequestContext[I, TRequest]]()
+        output_future = asyncio.create_task(
+            self._await_output(self._execute_request(call, request_future))
+        )
+        request_context = await request_future
+        input_stream = self.protocol.create_event_publisher(
```

Comment:
It doesn't look like we have a concrete implementation for either the publisher or receiver. The old workflow looks like it was calling execute on the operation directly from the `_duplex_stream` method. Is the intent we need to implement that for every protocol going forward?